### ISSUE TITLE
Middleware for nats.micro

### ISSUE THREAD

**Issue by kmpm** [nats-io/nats.go#1241]
## Feature Request
It seems as if there could be use for some kind of middleware when using nats.micro service.

#### Use Case:
Easliy add things to all endpoint handlers.
For example logging or header validation.

#### Proposed Change:
Some kind of middleware support. 
Preferably on root service level as well as group level.

#### Who Benefits From The Change(s)?
Everyone who has to do basic processing and validation for many endpoints.

#### Alternative Approaches
Nesting handlers.

**Comment by ripienaar** [comment]
I agree, however as this is quite low level and the purpose is to specify a common pattern and api for all our language its quite hard to specify middleware generically (think languages like C)

We intend to build language specific higher order libraries ontop of this - for example to run a proto file into a service using code generation - and there we have more flexibility to do language specific idiomatic things.

So in experiments I was able to design a http handler like middleware chain ontop of this in a way that's Go idiomatic without having to try come up with contrived language generic specifications to achieve that.
